File: 1106.解析布尔表达式.py
#
# @lc app=leetcode.cn id=1106 lang=python3
#
# [1106] 解析布尔表达式
#

# @lc code=start
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        st = []
        for c in expression:
            if c == ',': continue
            if c != ')':
                st.append(c)
                continue
            t = f = 0
            while st[-1] != '(':
                if st.pop() == 't': t += 1
                else: f += 1
            st.pop()
            op = st.pop()
            if op == '!':
                st.append('t' if f == 1 else 'f')
            elif op == '&':
                st.append('t' if f == 0 else 'f')
            elif op == '|':
                st.append('t' if t != 0 else 'f')
        return st[-1] == 't'
# @lc code=end

File: 1106.解析布尔表达式.cpp
/*
 * @lc app=leetcode.cn id=1106 lang=cpp
 *
 * [1106] 解析布尔表达式
 */
#include <bits/stdc++.h>
using namespace std;
// @lc code=start
class Solution {
public:
    bool parseBoolExpr(string expression) {
        stack<char> st;
        for (char c: expression) {
            if (c == ',') continue;
            if (c != ')') {
                st.push(c);
                continue;
            }
            int t = 0, f = 0;
            while (!st.empty() && st.top() != '(') {
                if (st.top() == 'f') {
                    f += 1;
                } else {
                    t += 1;
                }
                st.pop();
            }
            st.pop();
            char op = st.top(); st.pop();
            if (op == '!') {
                st.push(f == 1 ? 't' : 'f');
            } else if (op == '&') {
                st.push(f == 0 ? 't' : 'f');
            } else if (op == '|') {
                st.push(t != 0 ? 't' : 'f');
            }
        }
        return st.top() == 't';
    }
};
// @lc code=end

